src/auth/service.py

```diff
@@ -0,0 +1,143 @@
+
+    L.info("Successfully retrieved user info for subject %s", user_info.subject_id)
+    return user_info
+
+
+def validate_access_token(token: str, project_context: ProjectContext | None = None) -> TokenInfo:
+    """Validate the access token."""
+    token_info = get_token_info(token)
+    _validate_token_expiration_date(token_info)
+
+    user_info = get_user_info(token)
+    _validate_user_groups(user_info, project_context)
+
+    L.info("Token validated successfully for subject %s", user_info.subject_id)
+    return token_info
+
+
+def _validate_token_expiration_date(token_info):
+    """Validate that toke has not been expired."""
+    now = get_timestamp_now()
+    if token_info.expires_at < now:
+        message = f"Access token has expired.\nExpiration date: {token_info.expires_at}\nNow: {now}"
+        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail=message)
+
+
+def _validate_user_groups(user_info, project_context: ProjectContext | None) -> None:
+    """Validate that project context vlab/proj is in user groups."""
+    if project_context and not _context_in_groups(project_context, user_info.groups):
+        message = (
+            "Access token groups are not consistent with the project context.\n"
+            f"Project Context: {project_context}\n"
+            f"Subject ID     : {user_info.subject_id}\n"
+            f"User groups    : {user_info.groups}"
+        )
+        L.error(message)
+        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail=message)
+
+
+def _context_in_groups(context: ProjectContext, groups: list[str]) -> bool:
+    """Check if the context is in the groups."""
+    proj_ids = {g.split("/")[2] for g in groups if g.startswith("/proj")}
+    vlab_ids = {g.split("/")[2] for g in groups if g.startswith("/vlab")}
+    return str(context.virtual_lab_id) in vlab_ids and str(context.project_id) in proj_ids
+
+
+def validate_token_pair(
+    token_pair: TokenPairInput, project_context: ProjectContext | None = None
+) -> TokenPair:
+    """Validate the token pair."""
+    access_token_info = validate_access_token(
+        token=token_pair.access_token, project_context=project_context
+    )
+
+    refresh_token_info = get_token_info(token_pair.refresh_token)
+
+    if access_token_info.subject_id != refresh_token_info.subject_id:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Access token and refresh token subject IDs do not match.",
+        )
+
+    return TokenPair(
+        access_token=token_pair.access_token,
+        refresh_token=token_pair.refresh_token,
+        expires_at=access_token_info.expires_at,
+        subject_id=access_token_info.subject_id,
+    )
+
+
+def refresh_token_pair(token_pair: TokenPair) -> TokenPair:
+    """Refresh the access token."""
+    # TODO: check if access token is still valid within a margin of 1 minute
+
+    data = keycloak_openid.refresh_token(token_pair.refresh_token)
```

Comment:
@danifr a small question regarding the refresh tokens: do we have expiration time for them or any other limiting factor (e.g. max session time) that will render them invalid over time? (Not taking into account the revocation - that is very special case).

Reply:
Yeah the refresh token expires after 10h of being issued

Reply:
Can we increase that to 1 week for bbp-workflow?

Reply:
If you decode the refresh token you could access the `exp` claim.

<img width="531" alt="Screenshot 2025-03-07 at 10 53 40" src="https://github.com/user-attachments/assets/f4c6e530-6e86-4782-901a-7f1b6c3d7a38" />

```
{
  "exp": 1741377012,
  "iat": 1741341012,
  "jti": "4973c2ce-be61-4495-a191-ceeafecaa3b4",
  "iss": "https://www.openbraininstitute.org/auth/realms/SBO",
  "aud": "https://www.openbraininstitute.org/auth/realms/SBO",
  "sub": "172c3c0d-5e7d-4e98-9beb-58bc5fd1d2cd",
  "typ": "Refresh",
  "azp": "sbo-core-webapp",
  "sid": "822adbf3-3ed4-4056-8be7-10ebf0da4eec",
  "scope": "web-origins openid profile email roles groups basic acr"
}
```

Reply:
I see. The refresh_token I get with my local account and the bbp-workflow client with the password grant_type  is an offline token:
```
{
  "iat": 1741341711,
  "jti": "6928ca0d-cfab-4851-b1fd-d0b2b73ef796",
  "iss": "https://www.openbraininstitute.org/auth/realms/SBO",
  "aud": "https://www.openbraininstitute.org/auth/realms/SBO",
  "sub": "172c3c0d-5e7d-4e98-9beb-58bc5fd1d2cd",
  "typ": "Offline",
  "azp": "bbp-workflow",
  "sid": "0f12b54e-2e75-438d-922f-0e102fc6b0c6",
  "scope": "web-origins openid profile email roles groups offline_access basic"
}
```
Will that client always return offline tokens that don't expire. I am slightly confused :)

Reply:
that's because you are including the scope `offline_access`